### ISSUE TITLE
What prefixes should macros use for imports?

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/language#3500]
Currently, we just do something like `_prefix0` etc. We are probably going to move to `_$0` since it is shorter. But any scheme like this has the potential for problems, however rare they might be.

One possible option would be to reserve an entire range of identifier names, matching a pattern such as `_\$\d+`, as only being usable as import prefixes, and simply block their use for anything else. I have no idea how problematic that might be to roll out, but given it has a leading `_` it should always only be a local change. Maybe it is a horrible idea for other reasons though :).

cc @scheglov @munificent @davidmorgan

**Comment by jakemac53** [comment]
See related discord thread starting here https://discord.com/channels/608014603317936148/1169724112537657466/1182092062057828382

**Comment by lrhn** [comment]
Can't read the discord, but I wouldn't worry overly about this, and not make any language changes for it.

I'd just say that, fx, a fixed prefix+ `$` + global-unique-ID like `$macro_prefix$1x2z3` will be used, and if your code manages to contain that as an identifier too, then macros will fail for that library. So stop doing that.

You can't predict the ID. (Maybe you can, but you shouldn't. It could just be one global counter per compilation, or it could be randomized. Anything goes, as long as it's all fresh.)

Maybe just use `$macro_prefix` as the default and allow any macro implementation to choose its own prefix (but not ID), or allow a library that applies a macro to also annotate that macro application with a `@dart_macros_api.ImportPrefix(r"_my_own_unique_prefix")`, and then the macro API will recognize that and use that as (default) prefix for the macro implementation.

Hard-coding names into the language in order to avoid naming conflicts seems like a big hammer for not much nail.
Code generation has never really had big problems with creating fresh names, problems only start if one wants them to also be short.

**Comment by jakemac53** [comment]
These prefixes do appear a _lot_, so keeping them short does make the code significantly more readable.

I didn't think we would be willing to be OK with the possibility of a language feature creating conflicting names like this, but if we are then that SGTM, and I think even with the short names it is unlikely to actually conflict.

My biggest concern would be conflicts from other macro generated symbols, but we could publish the scheme being used and tell macros not to use the same scheme, or given them access to an API for generating names.

**Comment by rrousselGit** [comment]
> These prefixes do appear a lot, so keeping them short does make the code significantly more readable.

I'm not even sure about that

We lose a lot of readability by having an identifier we don't know about.

When folks use import prefixes, they generally use the package name as prefix. That's much more readable and mostly conflict proof

**Comment by munificent** [comment]
My understanding is that @scheglov is doing some work to have the macro generated code omit prefixes that aren't necessary to avoid collisions. Given that, if we're able to omit most prefixes, then I think using a verbose synthesized prefix name is less of an issue.

**Comment by scheglov** [comment]
It is mostly implemented.
For example
```dart
import 'json_serializable.dart';

@JsonSerializable()
class A {
  final int foo;
  final int bar;
}
```
produces
```dart
library augment 'package:test/test.dart';

import 'package:test/json_serializable.dart';
import 'dart:core';

augment class A {
  @FromJson()
  external A.fromJson(Map<String, Object?> json);
  @ToJson()
  external Map<String, Object?> toJson();
  augment A.fromJson(Map<String, Object?> json, )  : this.foo = json['foo'] as int,
this.bar = json['bar'] as int;
  augment Map<String, Object?> toJson() {
    var json = <String, Object?>{};
    json['foo'] = this.foo;
json['bar'] = this.bar;
    return json;
  }
}
```

I'd like to have empty lines between class members, and formatting, but this is a different story.

**Comment by jakemac53** [comment]
Yeah that is great, I think that @johnniwinther also implemented something to guarantee prefixes are unique.

**Comment by johnniwinther** [comment]
"guarantee" = if the macro doesn't split identifier across code parts, i.e. generating "ab" by generating "a" followed by generating "b".
